### PR TITLE
Fixed and refactored Optimizers

### PR DESCRIPTION
**Pull request by ferrine**
link #2164

### CODE REVIEW

**Comment by ferrine** [comment]
@twiecki, please review suggested API

File: pymc3/variational/updates.py [hunk 1]
```diff
@@ -451,6 +491,20 @@ def adagrad(loss_or_grads=None, params=None, learning_rate=1.0, epsilon=1e-6):
 
     .. [2] Chris Dyer:
            Notes on AdaGrad. http://www.ark.cs.cmu.edu/cdyer/adagrad.pdf
+    
+    Examples
+    --------
+    >>> a = theano.shared(1.)
+    >>> b = a*2
+    >>> updates = adagrad(b, [a], learning_rate=.01)
+    >>> isinstance(updates, dict)
```
Comment (twiecki) [inline pymc3/variational/updates.py:500]: Not sure the types are that useful to a user. I think how to use it practically or the actual outputs produced would be more valuable. Can't we set up a small optim problem?

Comment (ferrine) [reply pymc3/variational/updates.py:500]: Sure

Comment (ferrine) [reply pymc3/variational/updates.py:500]: To be honest I would rather change example in header.

Comment (twiecki) [reply pymc3/variational/updates.py:500]: How do you mean?

Comment (ferrine) [reply pymc3/variational/updates.py:500]: https://github.com/ferrine/pymc3/blob/6431bd578bdb14e701c129f403e6775218d44c99/pymc3/variational/updates.py#L77

Comment (twiecki) [reply pymc3/variational/updates.py:500]: OK, fair enough.

**Pull request merged**